runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoEvaluatorFactory.java

```diff
@@ -104,12 +104,12 @@ public void cleanup() throws Exception {
             .getExecutionContext(application, inputBundle.getKey())
             .getOrCreateStepContext(stepName, stepName);
 
-    DoFnLifecycleManager fnManager =
-        fnClones.getUnchecked(hooks.getDoFn(application.getTransform()));
+    DoFn<InputT, OutputT> doFn = hooks.getDoFn(application.getTransform());
+    DoFnLifecycleManager fnManager = fnClones.getUnchecked(doFn);
     try {
       return DoFnLifecycleManagerRemovingTransformEvaluator.wrapping(
           hooks.createParDoEvaluator(
-              evaluationContext, application, stepContext, (DoFn<InputT, OutputT>) fnManager.get()),
+              evaluationContext, application, stepContext, doFn),
```

Comment:
Ideally we have `DoFn<InputT, OutputT> DoFnInvoker#getFn()` which provides the original fn (for creating `ProcessContext` objects, and then within the runner we never pass `DoFn` anywhere (we wrap it in a `DoFnInvoker` whenever we deserialize it, and then only use the invoker if we would need to interact with it)

This might take some work on `DoFnRunners` as well